### PR TITLE
[Linux][LLDB][GDB/JIT] Add support for lldb 'breakpoint set' for jitted code

### PR DESCRIPTION
**Pull request by Dmitri-Botcharnikov**
This PR adds support for GDB/JIT feature ability to set breakpoints on managed code by using lldb 'breakpoint set' command:
```
(lldb) breakpoint set -f hello3.cs -l 8
Breakpoint 2: no locations (pending).
WARNING:  Unable to resolve breakpoint to any actual locations.
(lldb) c
Process 13004 resuming
1 location added to breakpoint 1
1 location added to breakpoint 2
Process 13004 stopped
* thread #1: tid = 13004, 0xaf9a6ffc JIT(0xeb960)`show_message + 44 at hello3.cs:8, name = 'corerun', stop reason = breakpoint 2.1
    frame #0: 0xaf9a6ffc JIT(0xeb960)`show_message + 44 at hello3.cs:8
   5        public class Program
   6        {
   7            public void show_message()
-> 8            {
   9                string a = "Hello World";
   10               string b = "CoreCLR!!!";
   11               Console.WriteLine(a + " " + b);
(lldb) n
Process 13004 stopped
* thread #1: tid = 13004, 0xaf9a6ffe JIT(0xeb960)`show_message + 46 at hello3.cs:9, name = 'corerun', stop reason = step over
    frame #0: 0xaf9a6ffe JIT(0xeb960)`show_message + 46 at hello3.cs:9
   6        {
   7            public void show_message()
   8            {
-> 9                string a = "Hello World";
   10               string b = "CoreCLR!!!";
   11               Console.WriteLine(a + " " + b);
   12           }
(lldb) c
```
This PR changes GDB/JIT feature so that debugging information is generated for all jitted methods in assemblies listed in `CORECLR_GDBJIT` environment variable:
```
export CORECLR_GDBJIT=hello3.1.exe
```
cc @lucenticus @chunseoklee @lemmaa @mikem8361 @jkotas

### CODE REVIEW

**Comment by Dmitri-Botcharnikov** [comment]
@ayuckhulk

File: src/vm/prestub.cpp [hunk 1]
```diff
@@ -224,13 +224,17 @@ void DACNotifyCompilationFinished(MethodDesc *methodDesc)
 
         _ASSERTE(modulePtr);
 
+#ifndef FEATURE_GDBJIT
```
Comment (mikem8361) [inline src/vm/prestub.cpp]: I don't understand why GDBJIT always needs this SOS notification or why would it need the notification at all?

File: src/vm/gdbjit.cpp [hunk 2]
```diff
@@ -287,17 +295,67 @@ void NotifyGdb::MethodCompiled(MethodDesc* MethodDescPtr)
...
+    if (length == 0)
+        return;
+    LPWSTR wszModuleFile = new (nothrow) WCHAR[length+1];
+    length = MultiByteToWideChar(CP_UTF8, 0, szModuleFile, -1, wszModuleFile, length);
+
+    if (length == 0)
+        return;
+
+    DWORD cCharsNeeded;
+    cCharsNeeded = GetEnvironmentVariableW(W("CORECLR_GDBJIT"), NULL, 0);
```
Comment (mikem8361) [inline src/vm/gdbjit.cpp]: It seems like a lot of unnecessary overhead to read the environment variable for ever method jit'ed.  Doesn't it only need to be read once and saved?

File: src/vm/gdbjit.cpp [hunk 3]
```diff
@@ -287,17 +295,67 @@ void NotifyGdb::MethodCompiled(MethodDesc* MethodDescPtr)
...
+    BOOL isUserDebug = FALSE;
+
+    if ((cCharsNeeded != 0) && (cCharsNeeded < MAX_LONGPATH))
+    {
+        LPWSTR wszModuleNames = new WCHAR[cCharsNeeded+1];
```
Comment (mikem8361) [inline src/vm/gdbjit.cpp]: You can use NewArrayHolder<WCHAR> wszModuleNames here and don't need to worry about delete'ing it at the end of the block.

File: src/vm/gdbjit.cpp [hunk 4]
```diff
@@ -287,17 +295,67 @@ void NotifyGdb::MethodCompiled(MethodDesc* MethodDescPtr)
...
+        cCharsNeeded = GetEnvironmentVariableW(W("CORECLR_GDBJIT"), wszModuleNames, cCharsNeeded);
+        if (cCharsNeeded != 0)
+        {
+            LPWSTR wszModuleName = new WCHAR[cCharsNeeded+1];
```
Comment (mikem8361) [inline src/vm/gdbjit.cpp]: NewArrayHolder can be used here too.

File: src/vm/gdbjit.cpp [hunk 5]
```diff
@@ -287,17 +295,67 @@ void NotifyGdb::MethodCompiled(MethodDesc* MethodDescPtr)
...
     const char *szModulePath, *szModuleFile;
-    
     SplitPathname(szModName, szModulePath, szModuleFile);
-    
+
+
+    int length = MultiByteToWideChar(CP_UTF8, 0, szModuleFile, -1, NULL, 0);
+    if (length == 0)
+        return;
+    LPWSTR wszModuleFile = new (nothrow) WCHAR[length+1];
```
Comment (mikem8361) [inline src/vm/gdbjit.cpp]: NewArrayHolder can be used here too.

File: src/vm/gdbjit.cpp [hunk 6]
```diff
@@ -947,6 +1094,19 @@ int NotifyGdb::Leb128Encode(int32_t num, char* buf, int size)
     return i;
 }
 
+#ifdef _DEBUG
```
Comment (mikem8361) [inline src/vm/gdbjit.cpp]: It looks like if you use FEATURE_GDBJIT with a DEBUG build a file will be written for each method jit'ed that is in the env var. It looks like this is only for your internal debugging use. Unless you really need it all the time, it should be ifdef'ed with something else.

**Comment by chunseoklee** [comment]
@Dmitri-Botcharnikov Did you run on arm32/linux ? I have failed to run on arm32/linux.

**Comment by Dmitri-Botcharnikov** [comment]
@chunseoklee : yes we checked it on linux/arm32. Could you please provide session log of failure?

File: src/vm/prestub.cpp (hunk 1 above)
Comment (Dmitri-Botcharnikov) [inline src/vm/prestub.cpp]: @mikem8361 the notification triggers generation of GDBJIT debugging info.

**Comment by chunseoklee** [comment]
@Dmitri-Botcharnikov we'd better discuss about arm32/linux on #6955

File: src/vm/prestub.cpp (hunk 1 above)
Comment (mikem8361) [inline src/vm/prestub.cpp]: Ok. Sounds good.

**Comment by Dmitri-Botcharnikov** [comment]
@mikem8361 , thank you for review.

**Comment by Dmitri-Botcharnikov** [comment]
@mikem8361 could this be merged?

**Pull request merged**